File: windows-app/Plantoir.Core/Models/CourseArchiver.cs
using System.Globalization;
using System.IO.Compression;

namespace Plantoir.Core.Models;

/// <summary>
/// Removes a course or section without destroying content: zip first into
/// courses/_backups/&lt;CODE&gt;/, then delete. The archive root contains the
/// folder itself (section3/…, ICS3U/…) — the restorer depends on that.
/// </summary>
public static class CourseArchiver
{
    /// <summary>The same exclusion list the setup wizard uses for its own backups.</summary>
    public static readonly IReadOnlyList<string> ExcludedFromArchives = new[]
    {
        ".merged_output", "node_modules", ".git", ".quartz-cache", ".cache",
        "dist", "build", "out", "__pycache__", ".DS_Store",
    };

    /// <summary>
    /// How many of the ASSISTANT's backups of one course are kept.
    ///
    /// A course full of images makes a large zip, and the assistant saves one
    /// per conversation whether or not anybody asked for it, so without a
    /// limit a term of chats fills a disk with copies of copies. (Until
    /// 2026-09-07 this comment described the mac and not this code, which
    /// saved one per CHANGE — so after six changes the copy from before the
    /// conversation had already been pruned. See
    /// <c>AssistWorkspace.BackUpOnceForThisConversation</c>.)
    ///
    /// A teacher's OWN backups are never counted here and never pruned. They
    /// made those on purpose; deciding on their behalf that a backup from
    /// last month has expired is not the app's call to make.
    /// </summary>
    public const int MostBackupsKept = 5;

    public static string TimestampedName(string prefix, DateTime stamp, string suffix = "") =>
        $"{prefix}_{stamp.ToString(ArchivedItem.StampFormat, CultureInfo.InvariantCulture)}{suffix}.zip";

    public static string BackupsDirectory(string coursesDirectory, string courseCode) =>
        Path.Combine(coursesDirectory, "_backups", courseCode);

    /// <summary>
    /// Saves a copy of an entire course — and touches nothing: the course
    /// stays exactly where it is. Made on purpose before risky editing so
    /// there is always a way back (row 106). The name
    /// (&lt;CODE&gt;_backup_&lt;timestamp&gt;.zip) is what separates a backup from an
    /// archive in the shared _backups folder.
    /// </summary>
    public static string BackUpCourse(Course course, string coursesDirectory, BackupMaker? maker = null)
    {
        var actualMaker = maker ?? BackupMaker.DefaultTeacher;
        string backupPath = Archive(course.DirectoryPath, course.Code + "_backup", coursesDirectory, course.Code, actualMaker.NameSuffix);
        PruneBackups(course.Code, coursesDirectory);
        return backupPath;
    }

    /// <summary>
    /// Deletes the oldest backups of one course until only
    /// <see cref="MostBackupsKept"/> are left.
    ///
    /// ONLY the assistant's own backups are pruned. A teacher's backup is a
    /// decision — deleting that on a schedule they never agreed to is the app
    /// overruling them about their own work.
    /// </summary>
    public static void PruneBackups(string courseCode, string coursesDirectory)
    {
        string backupsDir = BackupsDirectory(coursesDirectory, courseCode);
        if (!Directory.Exists(backupsDir)) return;

        var entries = Directory.GetFiles(backupsDir, "*.zip");
        var assistantBackups = new List<BackupItem>();
        foreach (var file in entries)
        {
            if (BackupItem.From(file, courseCode) is { } item && item.Maker is BackupMaker.Assistant)
            {
                assistantBackups.Add(item);
            }
        }

        if (assistantBackups.Count <= MostBackupsKept) return;

        // Newest first. Two backups made in the same second are ordered by name.
        assistantBackups.Sort((first, second) =>
        {
            if (first.BackedUpAt == second.BackedUpAt)
                return string.Compare(Path.GetFileName(second.FilePath), Path.GetFileName(first.FilePath), StringComparison.Ordinal);
            return second.BackedUpAt.CompareTo(first.BackedUpAt);
        });

        int kept = 0;
        foreach (var backup in assistantBackups)
        {
            kept++;
            if (kept > MostBackupsKept)
            {
                try { File.Delete(backup.FilePath); } catch { }
            }
        }
    }

    /// <summary>
    /// Writes an archive of an entire course folder WITHOUT removing it —
    /// for restores, which replace the course's contents in place so
    /// Obsidian's file watcher (anchored to the folder) keeps up.
    /// </summary>
    public static string ArchiveCourseWithoutRemoving(Course course, string coursesDirectory) =>
        Archive(course.DirectoryPath, course.Code, coursesDirectory, course.Code);

    /// <summary>
    /// Archive a whole course, then remove its folder — with the delete that
    /// survives readonly files and links (<see cref="CourseRestorer.DeleteTree"/>),
    /// because a build leaves both inside .merged_output.
    /// </summary>
    public static string ArchiveAndRemoveCourse(Course course, string coursesDirectory)
    {
        string archivePath = ArchiveCourseWithoutRemoving(course, coursesDirectory);
        CourseRestorer.DeleteTree(course.DirectoryPath);
        // A build outlives the content it was made from. Archive this course
        // and restore it next term and the notes that come back can be OLDER
        // than the site standing outside the working folder, so a freshness
        // check comparing dates says "already up to date" and the next publish
        // puts last term's pages online. The mac reaches the same rule a
        // different way - a build with no symlink pointing at it is cleared
        // rather than reused - and Windows has no link, so the clearing is
        // explicit at each moment a course's content is replaced.
        DiscardBuilds(coursesDirectory, course.Code);
        return archivePath;
    }

    /// <summary>
    /// Throw away a course's built site and build workspace, given the courses
    /// directory this course lives in.
    ///
    /// <para>The working folder is the courses directory's parent — the same
    /// relationship <c>Workspace.CoursesDirectory</c> creates going the other
    /// way. Best-effort inside <see cref="BuildOutputLocation"/>: a build
    /// still locked by a running preview must not turn "archive this course"
    /// into an error.</para>
    /// </summary>
    internal static void DiscardBuilds(string coursesDirectory, string courseCode, int? sectionNumber = null)
    {
        string? workingFolder = Path.GetDirectoryName(coursesDirectory.TrimEnd(
            Path.DirectorySeparatorChar, Path.AltDirectorySeparatorChar));
        if (string.IsNullOrEmpty(workingFolder)) return;
        string buildsRoot = BuildOutputLocation.BuildsRootFor(workingFolder!);
        if (sectionNumber is int number)
            BuildOutputLocation.DiscardBuildsFor(buildsRoot, courseCode, number);
        else
            BuildOutputLocation.DiscardBuildsFor(buildsRoot, courseCode);
    }

    /// <summary>
    /// Archive one section, remove it, and take its number out of the
    /// course's settings (num_sections follows along).
    /// </summary>
    public static string ArchiveAndRemoveSection(Course course, int sectionNumber, string coursesDirectory)
    {
        string sectionDir = course.SectionDirectory(sectionNumber);
        string archivePath = Archive(sectionDir, $"{course.Code}-section{sectionNumber}",
                                     coursesDirectory, course.Code);
        if (Directory.Exists(sectionDir)) CourseRestorer.DeleteTree(sectionDir);
        DiscardBuilds(coursesDirectory, course.Code, sectionNumber);
        // A scheduled deploy for a section that no longer exists cannot do
        // anything useful, and left alone it wakes up nightly to fail. Taking
        // the section's number out of the configuration is what makes the
        // launcher ask "Continue anyway?" about it, so this is also the other
        // half of the reason the wrapper runs non-interactively.
        Plantoir.Core.Assist.TaskScheduling.Cancel(
            Plantoir.Core.Assist.TaskScheduling.NameFor(course.Code, sectionNumber));
        var remaining = course.Configuration.SectionNumbers.Where(n => n != sectionNumber).ToList();
        course.Configuration.SetSectionNumbers(remaining);
        course.Configuration.Write(course.ConfigFilePath);
        return archivePath;
    }

    private static string Archive(string folderPath, string prefix, string coursesDirectory, string courseCode, string suffix = "")
    {
        string backupsDir = BackupsDirectory(coursesDirectory, courseCode);
        Directory.CreateDirectory(backupsDir);

        // Names are stamped to the second, and two operations can easily land
        // in the same one — an assistant publishing two classes in a row does
        // it every time. Without this the second backup throws, which (because
        // no backup means no edits) turns a routine sequence into a refusal.
        string archivePath = Path.Combine(backupsDir, TimestampedName(prefix, DateTime.Now, suffix));
        for (int attempt = 2; File.Exists(archivePath) && attempt < 100; attempt++)
            archivePath = Path.Combine(backupsDir,
                TimestampedName(prefix, DateTime.Now, suffix).Replace(".zip", $"-{attempt}.zip"));

        ZipFolder(folderPath, archivePath);
        return archivePath;
    }

    /// <summary>
    /// Zips a folder with the folder itself as the archive's root entry.
    /// The walk NEVER DESCENDS into an excluded folder — filtering paths
    /// after a full enumeration walked straight into .merged_output, where
    /// container-made links (content\Media) cannot be traversed by Windows
    /// and sank the whole backup. Reparse points are skipped for the same
    /// reason. A failure never leaves a partial zip behind.
    /// </summary>
    public static void ZipFolder(string folderPath, string archivePath)
    {
        string rootName = Path.GetFileName(folderPath.TrimEnd(Path.DirectorySeparatorChar));
        try
        {
            using var archive = ZipFile.Open(archivePath, ZipArchiveMode.Create);
            AddFolder(archive, folderPath, rootName);
        }
        catch
        {
            try { File.Delete(archivePath); } catch { }
            throw;
        }
    }

    private static void AddFolder(ZipArchive archive, string folderPath, string entryPrefix)
    {
        foreach (string file in Directory.EnumerateFiles(folderPath))
        {
            if (ExcludedFromArchives.Contains(Path.GetFileName(file))) continue;   // .DS_Store
            archive.CreateEntryFromFile(file, entryPrefix + "/" + Path.GetFileName(file),
                                        CompressionLevel.Optimal);
        }
        foreach (string sub in Directory.EnumerateDirectories(folderPath))
        {
            string name = Path.GetFileName(sub);
            if (ExcludedFromArchives.Contains(name)) continue;
            if ((new DirectoryInfo(sub).Attributes & FileAttributes.ReparsePoint) != 0) continue;
            AddFolder(archive, sub, entryPrefix + "/" + name);
        }
    }
}

File: mac-app/QuartzTeachers/Models/CourseArchiver.swift
import Foundation

/// Removes a course or one of its sections from the working folder —
/// but never destroys the content: the folder is zipped into
/// `courses/_backups/<CODE>/` first, matching the archive the setup
/// wizard writes before it changes a course.
enum CourseArchiver {

    // MARK: - Functions

    /// How many of the ASSISTANT's backups of one course are kept.
    ///
    /// A course full of images makes a large zip, and the assistant saves one
    /// per conversation whether or not anybody asked for it, so without a
    /// limit a term of chats fills a disk with copies of copies.
    ///
    /// A teacher's OWN backups are never counted here and never pruned. They
    /// made those on purpose; deciding on their behalf that a backup from
    /// last month has expired is not the app's call to make.
    static let mostBackupsKept: Int = 5

    /// Saves a copy of an entire course — and touches nothing: the course
    /// stays exactly where it is. Made on purpose before risky editing so
    /// there is always a way back. Returns the backup that was written.
    ///
    /// The name (`<CODE>_backup_<timestamp>.zip`) is what separates a
    /// backup from an archive in the shared `_backups` folder, and who made
    /// it rides in the same name — see `BackupMaker`.
    ///
    /// The oldest backups of this course are pruned afterwards, so the
    /// folder settles at `mostBackupsKept` instead of growing forever — with
    /// one exception, added with the calendar fix: a backup whose stamp
    /// cannot be true is neither counted nor deleted. See `pruneBackups`.
    @discardableResult
    static func backUpCourse(
        _ course: Course,
        coursesDirectoryURL: URL,
        madeBy maker: BackupMaker = .teacher
    ) throws -> URL {
        let backupURL: URL = try archive(
            folderURL: course.directoryURL,
            named: timestampedName(prefix: "\(course.code)_backup", suffix: maker.nameSuffix),
            forCourseCode: course.code,
            coursesDirectoryURL: coursesDirectoryURL
        )
        pruneBackups(forCourseCode: course.code, coursesDirectoryURL: coursesDirectoryURL)
        return backupURL
    }

    /// Deletes the oldest backups of one course until only
    /// `mostBackupsKept` are left.
    ///
    /// Only backups: archives (`<CODE>_<timestamp>.zip`) and the setup
    /// wizard's automatic zips (`<timestamp>.zip`) share this folder, and
    /// deleting one of those would throw away the only copy of something a
    /// teacher deliberately put away. `BackupItem.from` accepts nothing but
    /// this convention's own names, so asking it is the whole safeguard.
    static func pruneBackups(forCourseCode courseCode: String, coursesDirectoryURL: URL) {
        let backupsURL: URL = coursesDirectoryURL
            .appendingPathComponent("_backups")
            .appendingPathComponent(courseCode)
        guard let contents = try? FileManager.default.contentsOfDirectory(
            at: backupsURL,
            includingPropertiesForKeys: nil,
            options: [.skipsHiddenFiles]
        ) else {
            return
        }

        // ONLY the assistant's own backups are pruned.
        //
        // A teacher's backup is a decision — they pressed Back Up because they
        // were about to do something they were unsure of, or the app saved one
        // on their behalf before adding a section. Deleting that on a schedule
        // they never agreed to is the app overruling them about their own
        // work. The assistant's are different in kind: it makes one per
        // conversation whether or not anybody asked, so it is the assistant's
        // job to clear up after itself.
        //
        // A consequence worth knowing: the five kept are five ASSISTANT
        // backups. A teacher with twenty of their own keeps all twenty, and
        // they do not crowd out the assistant's five.
        var backups: [BackupItem] = []
        for fileURL in contents {
            guard let backup = BackupItem.from(fileURL: fileURL, courseCode: courseCode) else {
                continue
            }
            guard case .assistant = backup.maker else {
                continue
            }
            // A stamp that cannot be true is not allowed to decide what gets
            // DELETED. `ArchiveStamp` still reads such a name — a zip carried
            // from a Mac whose calendar this one does not use, say — so the
            // teacher can see it and restore it; but its date is the one
            // thing about it that is known to be wrong, and this list is
            // sorted by date before the tail of it is thrown away. A copy
            // stamped 2569 would sort as the NEWEST thing in the folder and
            // quietly take a real backup's place among the five that are
            // kept. Left out, it is never counted and never deleted.
            guard ArchiveStamp.couldHaveBeenStamped(backup.backedUpAt) else {
                continue
            }
            backups.append(backup)
        }
        if backups.count <= mostBackupsKept {
            return
        }

        // Newest first. Two backups made in the same second are ordered by
        // name, so the answer is the same every time it is asked.
        backups.sort { first, second in
            if first.backedUpAt == second.backedUpAt {
                return first.fileURL.lastPathComponent > second.fileURL.lastPathComponent
            }
            return first.backedUpAt > second.backedUpAt
        }

        var keptSoFar: Int = 0
        for backup in backups {
            keptSoFar += 1
            if keptSoFar > mostBackupsKept {
                try? FileManager.default.removeItem(at: backup.fileURL)
            }
        }
    }

    /// Writes an archive of an entire course folder without removing it —
    /// for restores, which replace the course's CONTENTS in place so
    /// Obsidian's file watcher (anchored to the folder) keeps up.
    /// Returns the archive that was written.
    @discardableResult
    static func archiveCourse(_ course: Course, coursesDirectoryURL: URL) throws -> URL {
        return try archive(
            folderURL: course.directoryURL,
            named: timestampedName(prefix: course.code),
            forCourseCode: course.code,
            coursesDirectoryURL: coursesDirectoryURL
        )
    }

    /// Archives and removes an entire course folder.
    /// Returns the archive that was written.
    @discardableResult
    static func archiveAndRemoveCourse(_ course: Course, coursesDirectoryURL: URL) throws -> URL {
        let archiveURL: URL = try archiveCourse(course, coursesDirectoryURL: coursesDirectoryURL)
        try FileManager.default.removeItem(at: course.directoryURL)
        // The built website lives OUTSIDE the working folder now, so removing
        // the course folder no longer removes it — see `BuildOutputLocation`.
        // Two reasons it has to go with the course: it is otherwise invisible
        // litter in Application Support that nothing will ever name again, and
        // a course of the same code restored later would find a built site
        // older than its own pages and be told it was up to date.
        BuildOutputLocation.discardBuild(
            forWorkingFolder: coursesDirectoryURL.deletingLastPathComponent(),
            courseCode: course.code
        )
        return archiveURL
    }

    /// Archives and removes one section folder, and takes that section
    /// out of the course's saved settings so it stops being listed.
    @discardableResult
    static func archiveAndRemoveSection(
        _ sectionNumber: Int,
        from course: Course,
        coursesDirectoryURL: URL
    ) throws -> URL {
        let sectionURL: URL = course.sectionDirectoryURL(forSection: sectionNumber)
        let archiveURL: URL = try archive(
            folderURL: sectionURL,
            named: timestampedName(prefix: "\(course.code)-section\(sectionNumber)"),
            forCourseCode: course.code,
            coursesDirectoryURL: coursesDirectoryURL
        )
        if FileManager.default.fileExists(atPath: sectionURL.path) {
            try FileManager.default.removeItem(at: sectionURL)
        }
        // This section's built website goes with it, for the same reason the
        // whole course's does: a section restored later carries the timestamps
        // it had when it was archived, which can be OLDER than the site built
        // from it — so a build left standing would read as up to date and
        // publish pages the teacher had already replaced.
        BuildOutputLocation.discardSectionBuild(
            forWorkingFolder: coursesDirectoryURL.deletingLastPathComponent(),
            courseCode: course.code,
            sectionNumber: sectionNumber
        )

        var remainingSections: [Int] = []
        for existingSection in course.sectionNumbers {
            if existingSection != sectionNumber {
                remainingSections.append(existingSection)
            }
        }
        course.configuration.setSectionNumbers(remainingSections)
        try course.configuration.write(to: course.configFileURL)

        return archiveURL
    }

    /// Things that are rebuilt rather than written, and so are left out of
    /// an archive. `.merged_output` is the important one: it holds a whole
    /// Quartz checkout with its dependencies, which dwarfs the course a
    /// teacher actually wrote and can always be produced again.
    ///
    /// The same list the setup wizard uses for its own backups.
    static let excludedFromArchives: [String] = [
        ".merged_output",
        "node_modules",
        ".git",
        ".quartz-cache",
        ".cache",
        "dist",
        "build",
        "out",
        "__pycache__",
        ".DS_Store",
    ]

    /// Zips a folder into `courses/_backups/<CODE>/<name>.zip`.
    private static func archive(
        folderURL: URL,
        named archiveName: String,
        forCourseCode courseCode: String,
        coursesDirectoryURL: URL
    ) throws -> URL {
        let backupsURL: URL = coursesDirectoryURL
            .appendingPathComponent("_backups")
            .appendingPathComponent(courseCode)
        try FileManager.default.createDirectory(at: backupsURL, withIntermediateDirectories: true)
        let archiveURL: URL = backupsURL.appendingPathComponent(archiveName)

        // Zipped with the system's own tool rather than NSFileCoordinator,
        // because this one can leave things out — and a course's built
        // output is many times the size of the course itself.
        var arguments: [String] = ["-r", "-q", "-X", archiveURL.path, folderURL.lastPathComponent]
        arguments.append("-x")
        for name in excludedFromArchives {
            arguments.append("*/\(name)/*")
            arguments.append("*/\(name)")
        }

        let zipper: Process = Process()
        zipper.executableURL = URL(fileURLWithPath: "/usr/bin/zip")
        zipper.arguments = arguments
        zipper.currentDirectoryURL = folderURL.deletingLastPathComponent()
        let errors: Pipe = Pipe()
        zipper.standardError = errors
        try zipper.run()
        zipper.waitUntilExit()
        if zipper.terminationStatus != 0 {
            let data: Data = errors.fileHandleForReading.readDataToEndOfFile()
            let reason: String = String(data: data, encoding: .utf8) ?? "unknown error"
            throw NSError(
                domain: "CourseArchiver",
                code: Int(zipper.terminationStatus),
                userInfo: [NSLocalizedDescriptionKey: "Could not write the archive: \(reason.trimmingCharacters(in: .whitespacesAndNewlines))"]
            )
        }
        return archiveURL
    }

    /// "ICS3U_2026-08-09_141530.zip", or with a suffix,
    /// "ICS3U_backup_2026-08-09_141530_assistant-section1.zip".
    ///
    /// The moment is spelled by `ArchiveStamp`, which is also what reads it
    /// back — a writer with a formatter of its own is how the two came to
    /// disagree in the first place.
    private static func timestampedName(prefix: String, suffix: String = "") -> String {
        return "\(prefix)_\(ArchiveStamp.text(for: Date()))\(suffix).zip"
    }
}
